### ISSUE TITLE
Make private URL distribution e-mails customisable

### ISSUE THREAD

**Issue by czlee** [TabbycatDebate/tabbycat#531]
A nice-to-have, allow private URL distribution e-mails to be edited before sending.

- Should be considered an advanced feature; i.e., it shouldn't be part of the standard workflow.
- Should include default text (basically the string we currently use).
- Validation should check that all merge fields are present (to allow for typos).
- Should show a completed example, or even better allow you to switch between all of them like in mail merge (but the latter is probably a bit overkill). Actually, this part probably should be part of the standard workflow.

**Comment by tienne-B** [comment]
Is there a point that I didn't address here? Also, I assume that adding the entry to the CHANGELOG (if these patches merit) will be handled by you guys.

**Comment by czlee** [comment]
Sorry, I had to head out to something so ran out of time to post a message here. The PR is great, there are just a few tweaks I'd like on it. Do you want to have a crack at them? @philipbelesky and I are really happy with your work so far and are happy to do these ourselves, but I wanted to give you first dibs.

- I'd use `<NAME>`, `<TEAM>` and `<URL>` (in all caps) rather than `%name`, `%team` and `%url`. If the angle brackets get in the way of HTML stuff, then `[[NAME]]` (that's two sets of brackets), _etc._
- I'd like users to be at least warned if `<NAME>` and `<TEAM>` aren't present. The purpose isn't so much to prevent users from excluding the fields, as it is to catch typos in those fields (so _e.g._ `<NAM>` doesn't go unsubstituted, or to catch out `[NAME]` if double square brackets are used).
- The `help_text` doesn't currently display in the form; it'd be nice if it did.
- It'd be nice to save the text that was last used to blast e-mails, as the initial value next time the form is used (rather than always reverting to the same default), so that if subsequent emails need to be sent out for new participants, they will use the same text. This could be done using the dynamic preferences system—in its own section; I don't think there's a need to make this modifiable from the settings GUI.

There are a few further enhancements I have in mind as well, but they can go in a successor issue once this one is closed 🙂

**Comment by tienne-B** [comment]
(1) **Done.**
(2) Could be done by adding a hidden field which if set would ignore the `ValidationError` that parameters like `<NAME>` is absent, but can't seem to set it after having the error. Also, unsure of how well this approach would go with point 4, as it isn't a "preference" **I'll leave this point.**
(3) is a Django problem; seems that `help_text` can't be set in the class (like `label`). I've put the text in its instances (and added the `<TEAM>` parameter where applicable) **Done.**
(4) With adding all the fields as preferences, the forms would be converted to `dynamic_preferences.forms`. I wouldn't know how to have translatable default values for preferences with variable substitution (for the tournament name--don't want to increase the number of `replace`s). **I'll leave this point.**

PR #642 created.

**Comment by czlee** [comment]
Sounds good! Just a small regression to clear up on #642, then we'll merge it in and we'll do the rest. Thanks again!

**Comment by tienne-B** [comment]
For (2) and (4), we could just put the subject/message in the preferences. In fact get rid of the intermediate interface before sending messages, just have a `POST` button, which is disabled if all emails were sent. The (correct) expectation is that only people with changed data will be emailed. As it is important to know who got the messages (or who _didn't_), there could be a warning message after the success message stating who did not get sent a message. I think that would be the simplest and most best solution to standardize these email actions and get rid of this task.

**Comment by czlee** [comment]
For what it's worth, preferences don't _have_ to show up in the preference forms. You can just manipulate them directly: [django-dynamic-preferences documentation](https://django-dynamic-preferences.readthedocs.io/en/latest/quickstart.html#retrieve-and-update-preferences). (For the tournament preferences, `preferences` is an attribute of each `Tournament` object. I know we use a method `Tournament.pref()`, that's mostly an extra layer of caching, but it also allows us to omit the section name.) I know we don't really use this pattern much in our code (because we assume all preferences go via the forms), but if we need to here, you should feel at liberty to.

**Comment by tienne-B** [comment]
What would be really good is if there could be translations of the default, so that the default text does not have to be English. Had to get rid of that to use dynamic-preferences. My idea is to basically ignore (2) and standardize the email forms to all be handled in the preferences. With that, we have one fewer view as well.

**Comment by czlee** [comment]
Yeah, I agree that'd be nice. The consequential question is then not whether to use django-dynamic-preferences, but how such a default would even function. Independent of implementation: The default is what gets used if it hasn't yet been set by the user. But if we have a language-dependent default, when does it become the actual setting, and which language is used?

(a) Do we have some conception of a "language of the tournament", explicitly set by the tab director?
(b) Do we set the default on tournament creation, according to whatever language the tournament creation form was presented in?
(c) Do we choose the language of whoever loads the emails form first, so that the second person to look at the form sees it in the language of the first user? 
(d) Do we save (and hence fix the language) _only_ when the first emails are actually _sent_, always loading the language of whoever's looking at the form _until it's first sent_, at which point that text (in whatever language, or gibberish, since the user can change it) becomes the setting?
(e) Do we save (and hence fix the language) _only_ if the user changed the default suggestion in some way?

(d) is the most complicated but I can see a case for why it might be the most "natural". It might not be as bad as it sounds: You make the preference default an empty string, and then have the emails _form_ (not the preference system) populate the message field with the _actual_ (translated) default _if the preference is empty_ (and use whatever's in the preference if it's not empty).

On another note—I got confused about how (2) relates to this. I thought (2) was about warning users if `<NAME>` and `<TEAM>` aren't present, which I thought was independent of (and not affected by) all of this?
